simpleflow/swf/executor.py

```diff
@@ -1167,3 +1189,30 @@ def get_event_details(self, event_type, event_name):
             raise ValueError('Unimplemented type {!r} for get_event_details'.format(
                 event_type
             ))
+
+    def handle_cancel_requested(self):
+        decision = swf.models.decision.WorkflowExecutionDecision()
+        is_current_decision = self._history.completed_decision_id < self._history.cancel_requested_id
+        should_cancel = self._workflow.should_cancel(self._history)
+        if not should_cancel:
+            return None  # ignore cancel
+        if is_current_decision:
+            self.on_canceled()
+            decision.cancel()
+            return [decision]
+        if self._history.cancel_failed:
+            logger.warning('failed: %s', self._history.cancel_failed)
+        if self._history.cancel_failed and self._history.cancel_failed_decision_task_completed_event_id == self._history.completed_decision_id:
+            # Per http://docs.aws.amazon.com/amazonswf/latest/apireference/API_Decision.html,
+            # we should call RespondDecisionTaskCompleted without any decisions; however this hangs the workflow...
```

Comment:
Why would a cancel failed? Looking at: http://docs.aws.amazon.com/amazonswf/latest/apireference/API_CancelWorkflowExecutionFailedEventAttributes.html => I'm not much afraid of OPERATION_NOT_PERMITTED in the general case, but UNHANDLED_DECISION? Do you know why this could happen?